File: backend/app/api/jobs.py
"""Job API routes."""

from __future__ import annotations

from dataclasses import asdict

from fastapi import APIRouter, Depends, HTTPException, Query

from app.api.papers import envelope
from app.schemas.papers import APIEnvelope, JobResponse
from core.services.papers import (
    JobCancelNotAllowedError,
    JobListInput,
    JobNotFoundError,
)
from core.services.papers.models import JobRecord
from core.services.papers.service import PaperService


router = APIRouter(prefix="/api/v1", tags=["jobs"])


def get_paper_service() -> PaperService:
    return PaperService()


def to_job_response(record: JobRecord) -> JobResponse:
    return JobResponse.model_validate(asdict(record))


def raise_http_error(exc: Exception) -> None:
    if isinstance(exc, JobNotFoundError):
        raise HTTPException(
            status_code=404,
            detail={"code": exc.code, "message": str(exc)},
        )
    if isinstance(exc, JobCancelNotAllowedError):
        raise HTTPException(
            status_code=409,
            detail={"code": exc.code, "message": str(exc)},
        )
    raise exc


@router.get("/jobs", response_model=APIEnvelope)
def list_jobs(
    page: int = 1,
    page_size: int = 20,
    resource_type: str | None = None,
    resource_id: int | None = None,
    job_status: str | None = Query(default=None, alias="status"),
    service: PaperService = Depends(get_paper_service),
) -> APIEnvelope:
    jobs, total = service.list_jobs(
        JobListInput(
            page=page,
            page_size=page_size,
            resource_type=resource_type,
            resource_id=resource_id,
            status=job_status,
        )
    )
    return envelope(
        [JobResponse.model_validate(asdict(job)) for job in jobs],
        meta={"page": page, "page_size": page_size, "total": total},
    )


@router.get("/jobs/{job_id}", response_model=APIEnvelope)
def get_job(
    job_id: str,
    service: PaperService = Depends(get_paper_service),
) -> APIEnvelope:
    try:
        return envelope(to_job_response(service.get_job(job_id)))
    except Exception as exc:  # pragma: no cover - centralized mapping
        raise_http_error(exc)
        raise


@router.post("/jobs/{job_id}/cancel", response_model=APIEnvelope)
def cancel_job(
    job_id: str,
    service: PaperService = Depends(get_paper_service),
) -> APIEnvelope:
    try:
        return envelope(to_job_response(service.cancel_job(job_id)))
    except Exception as exc:  # pragma: no cover - centralized mapping
        raise_http_error(exc)
        raise

File: frontend/src/lib/jobs.ts
import { request } from "@/lib/http";

export type JobRecord = {
  job_id: string;
  type: string;
  status: string;
  progress: number;
  message: string;
  resource_type: string;
  resource_id: number;
  resource_label: string;
  created_at: string;
  updated_at: string;
  result: Record<string, unknown> | null;
  error: {
    code: string;
    message: string;
    details?: Record<string, unknown>;
  } | null;
};

export type JobListResponse = {
  jobs: JobRecord[];
  total: number;
  page: number;
  pageSize: number;
};

export async function listJobs(input: {
  page?: number;
  pageSize?: number;
  resourceType?: string;
  resourceId?: number;
  status?: string;
} = {}): Promise<JobListResponse> {
  const params = new URLSearchParams();
  params.set("page", String(input.page ?? 1));
  params.set("page_size", String(input.pageSize ?? 10));
  if (input.resourceType) {
    params.set("resource_type", input.resourceType);
  }
  if (input.resourceId !== undefined) {
    params.set("resource_id", String(input.resourceId));
  }
  if (input.status) {
    params.set("status", input.status);
  }
  const envelope = await request<JobRecord[]>(`/api/v1/jobs?${params.toString()}`);
  return {
    jobs: envelope.data,
    total: Number(envelope.meta.total ?? envelope.data.length),
    page: Number(envelope.meta.page ?? input.page ?? 1),
    pageSize: Number(envelope.meta.page_size ?? input.pageSize ?? 10),
  };
}

export async function getJob(jobId: string): Promise<JobRecord> {
  const envelope = await request<JobRecord>(`/api/v1/jobs/${jobId}`);
  return envelope.data;
}

export async function cancelJob(jobId: string): Promise<JobRecord> {
  const envelope = await request<JobRecord>(`/api/v1/jobs/${jobId}/cancel`, {
    method: "POST",
  });
  return envelope.data;
}
